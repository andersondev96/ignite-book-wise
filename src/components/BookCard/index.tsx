import { useState } from 'react'
import dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime'
import 'dayjs/locale/pt-br'

import {
  AuthorInfo,
  AuthorSection,
  Avatar,
  BookCardContainer,
  BookInfoContent,
  BookInfoSection,
  Text,
} from './styles'
import { Stars } from '../Stars'

dayjs.extend(relativeTime)

interface Rating {
  id: string
  rate: number
  description: string
  created_at: string
  book_id: string
  user_id: string
  book: {
    id: string
    name: string
    author: string
    summary: string
    cover_url: string
    created_at: string
  }
  user: {
    id: string
    name: string
    avatar_url: string
  }
}

interface BookCardProps {
  rating: Rating
}

export const BookCard = ({ rating }: BookCardProps) => {
  const [expanded, setExpanded] = useState(false)

  const safeDescription = rating.description ?? ''

  return (
    <BookCardContainer>
      <AuthorSection>
        {rating.user && (
          <AuthorInfo href={`/profile/${rating.user.id}`}>
            <Avatar>
              <img src={rating.user.avatar_url} alt={rating.user.name} />
            </Avatar>
            <div>
              <span>{rating.user.name}</span>
              <p>{dayjs().locale('pt-br').from(dayjs(rating.created_at))}</p>
            </div>
          </AuthorInfo>
        )}
        <Stars rate={rating.rate} />
      </AuthorSection>
      <BookInfoSection>
        <img
          src={rating.book.cover_url.replace('public', '')}
          alt={rating.book.name}
        />
        <BookInfoContent>
          <strong>{rating.book.name}</strong>
          <span>{rating.book.author}</span>

          <Text>
            {safeDescription && typeof safeDescription === 'string' ? (
              expanded || safeDescription.length <= 100 ? (
                safeDescription
              ) : (
                <>
                  {safeDescription.slice(0, 100) + '...'}{' '}
                  <span onClick={() => setExpanded(true)}>ver mais</span>
                </>
              )
            ) : (
              ''
            )}
          </Text>
        </BookInfoContent>
      </BookInfoSection>
    </BookCardContainer>
  )
}
