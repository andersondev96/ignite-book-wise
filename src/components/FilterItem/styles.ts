import { styled } from '@/stitches.config'

export const Container = styled('div', {
  display: 'inline-flex',
  alignItems: 'center',
  justifyContent: 'center',

  height: '36px',
  padding: '0 20px',
  borderRadius: '999px',
  border: '1px solid $purple100',
  backgroundColor: 'transparent',

  fontSize: '$md',
  fontWeight: 500,
  color: '$purple100',
  whiteSpace: 'nowrap',

  cursor: 'pointer',
  userSelect: 'none',
  transition: 'all 0.2s ease',

  '&:hover': {
    background: '$purple200',
    color: '$white',
  },

  '&:focus': {
    outline: '2px solid $purple200',
    outlineOffset: '2px',
  },

  variants: {
    active: {
      true: {
        background: '$purple200',
        border: 'none',
        color: '$white',
      },
    },
  },
})
