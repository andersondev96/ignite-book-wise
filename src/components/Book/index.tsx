import * as Dialog from '@radix-ui/react-dialog'
import { Stars } from '../Stars'
import { BookInfo, BookName, Container } from './styles'
import { ModalBookDetails } from '../ui/ModalBookDetails'

interface BookSchema {
  id: string
  name: string
  author: string
  cover_url: string
  avgRating: number
}

interface BookProps {
  book: BookSchema
}

export const Book = ({ book }: BookProps) => {
  return (
    <Dialog.Root>
      <Dialog.Trigger asChild>
        <Container>
          <img src={book.cover_url.replace('public', '')} alt={book.name} />
          <BookInfo>
            <BookName>
              <strong>{book.name}</strong>
              <span>{book.author}</span>
            </BookName>
            <Stars rate={book.avgRating} />
          </BookInfo>
        </Container>
      </Dialog.Trigger>
      <ModalBookDetails id={book.id} />
    </Dialog.Root>
  )
}
