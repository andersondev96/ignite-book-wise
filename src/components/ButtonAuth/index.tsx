import type { ButtonHTMLAttributes, ReactNode } from 'react'
import { Container } from './styles'

type ButtonAuthProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  icon?: ReactNode
  image_url?: string
  title: string
}

export const ButtonAuth = ({
  icon,
  image_url,
  title,
  ...rest
}: ButtonAuthProps) => {
  return (
    <Container {...rest}>
      {icon || <img src={image_url} alt="" />}
      <span>{title}</span>
    </Container>
  )
}
