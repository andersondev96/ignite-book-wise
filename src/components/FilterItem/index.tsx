import { useCallback } from 'react'

import { Container } from './styles'

type FilterItemProps = {
  title: string
  active: boolean
  onClick: () => void
}

export const FilterItem = ({ title, active, onClick }: FilterItemProps) => {
  const handleClick = useCallback(() => {
    onClick()
  }, [onClick])

  return (
    <Container onClick={handleClick} isActive={active} tabIndex={0}>
      <span>{title}</span>
    </Container>
  )
}
