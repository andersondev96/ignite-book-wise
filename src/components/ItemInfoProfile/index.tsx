import type { ReactNode } from 'react'
import { Container, ItemInfo } from './styles'

interface ItemInfoProfileProps {
  icon: ReactNode
  title: string
  description: string
}

export const ItemInfoProfile = ({
  icon,
  title,
  description,
}: ItemInfoProfileProps) => {
  return (
    <Container>
      {icon}
      <ItemInfo>
        <strong>{title}</strong>
        <span>{description}</span>
      </ItemInfo>
    </Container>
  )
}
