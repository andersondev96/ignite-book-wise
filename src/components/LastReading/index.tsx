import { Container, Title } from './styles'
import { CaretRight } from '@phosphor-icons/react'
import { BookCard } from '../BookCard'
import { useEffect, useState } from 'react'
import { api } from '@/src/lib/axios'
import { useSession } from 'next-auth/react'

interface Book {
  id: string
  name: string
  author: string
  summary: string
  cover_url: string
  ratings: Array<{
    id: string
    rate: number
    created_at: string
    user: {
      id: string
      name: string
      avatar_url: string
    }
  }>
}

export const LastReading = () => {
  const [book, setBook] = useState<Book>()
  const { data: session } = useSession()
  const userId = session ? session.user.id : null

  useEffect(() => {
    if (userId) {
      api
        .get(`/ratings/last-reading?userId=${userId}`)
        .then((response) => {
          setBook(response.data)
        })
        .catch((error) => {
          console.log(error)
        })
    }
  }, [userId])

  book ? (
    <Container>
      <Title>
        <span>Sua Ãºltima leitura</span>
        <a href="#">
          Ver todas
          <CaretRight size={16} />
        </a>
      </Title>
      <BookCard
        name={book.ratings[0].user.name}
        avatar_url={book.ratings[0].user.avatar_url}
        rating_date={book.ratings[0].created_at}
        rate={book.ratings[0].rate}
        book_cover_url={book.cover_url.replace('public', '')}
        book_name={book.name}
        book_author={book.author}
        book_description={book.summary}
      />
    </Container>
  ) : (
    <></>
  )
}
