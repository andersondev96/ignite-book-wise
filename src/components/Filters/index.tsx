import { useCategory } from '@/src/contexts/CategoryContext'
import { FilterItem } from '../FilterItem'
import { Container } from './styles'
import { useCategories } from '@/src/hooks/useCategories'

export const Filters = () => {
  const { categories } = useCategories()
  const { setSelectedCategory } = useCategory()

  return (
    <Container>
      <FilterItem
        id=""
        title="Tudo"
        handleCategory={() => setSelectedCategory('')}
      />
      {categories.map((category) => {
        return (
          <FilterItem
            key={category.id}
            id={category.id}
            title={category.name}
            handleCategory={() => setSelectedCategory(category.id)}
          />
        )
      })}
    </Container>
  )
}
