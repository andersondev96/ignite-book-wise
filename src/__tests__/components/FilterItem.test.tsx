import { render, screen } from '@testing-library/react'
import { expect } from 'vitest'

import { FilterItem } from '@/src/components/FilterItem'

describe('<FilterItem />', () => {
  // Teste de snapshot
  it('renders with default styles when not active', () => {
    const { container } = render(
      <FilterItem title="suspense" active={false} onClick={() => {}} />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with active styles when active', () => {
    const { container } = render(
      <FilterItem title="suspense" active onClick={() => {}} />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders filter title', () => {
    const handleClick = vi.fn()

    render(<FilterItem title="suspense" active onClick={handleClick} />)

    expect(screen.getByText('suspense')).toBeInTheDocument()
  })

  it('calls onClick when clicked', () => {
    const handleClick = vi.fn()

    render(<FilterItem title="suspense" active={false} onClick={handleClick} />)

    const element = screen.getByText('suspense').parentElement!
    element.click()

    expect(handleClick).toHaveBeenCalled()
  })
})
