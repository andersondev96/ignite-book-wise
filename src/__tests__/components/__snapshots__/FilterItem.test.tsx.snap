// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<FilterItem /> > renders with active styles when active 1`] = `
<div
  class="c-dPKmXm c-dPKmXm-bWxTnJ-isActive-true"
  tabindex="0"
>
  <span>
    suspense
  </span>
</div>
`;

exports[`<FilterItem /> > renders with default styles when not active 1`] = `
<div
  class="c-dPKmXm"
  tabindex="0"
>
  <span>
    suspense
  </span>
</div>
`;
